README.md
pymad - a Python wrapper for the MPEG Audio Decoder library
===========================================================

pymad is a Python module that allows Python programs to use the MPEG Audio Decoder library. pymad provides a high-level API, similar to the pyogg module, which makes reading PCM data from MPEG audio streams a piece of cake.

MAD is available at http://www.mars.org/home/rob/proj/mpeg/

Access this module via `import mad`.  To decode
an mp3 stream, you'll want to create a `mad.MadFile` object and read data from
that.  You can then write the data to a sound device.  See the example
program in `test/` for a simple mp3 player that uses the `python-pyao` wrapper around libao for the sound
device.

pymad wrapper isn't as low level as the C MAD API is, for example, you don't
have to concern yourself with fixed point conversion -- this was done to
make pymad easy to use.

```python
import sys

import ao
import mad

mf = mad.MadFile(sys.argv[1])
dev = ao.AudioDevice(0, rate=mf.samplerate())
while 1:
    buf = mf.read()
    if buf is None:  # eof
        break
    dev.play(buf, len(buf))
```


To build, you need the distutils package, availible from
http://www.python.org/sigs/distutils-sig/download.html (it comes with
Python 2.0). Run "python setup.py build" to build and then as root run
"python setup.py install".  You may need to run the config_unix.py
script, passing it a --prefix value if you've installed your mad stuff
someplace weird.  Alternately, you can just create a file called
"Setup" and put in values for mad_include_dir, mad_lib_dir, and
mad_libs.  The file format for Setup is:

key = value

with one pair per line.

```shell
# python config_unix.py --prefix /usr/local
# python setup.py build
# python setup.py install --prefix /usr/local
```

Remember to make sure `/usr/local/python/site-packages/` is in your Python search path.


NEWS
NEWS for pymad
==============

0.8
---

2015-03-03

 * Reinitialize buffers after seeks to prevent decoding and output of
   partial content prior to the seek.  (Markus Demleitner)

0.6
---

2007-01-22

 * Fix decoding error with files containing largish id3v2 tags (Joerg
   Lehmann)
 * Fix length calculation (Richard Adenling)

0.5.1 (07/10/03)
-----

* Fix segfault when cleaning up.

0.5 (01/10/03)
---

* Now can read from any file-like Python object.
* Added example program for playing internet radio streams.

0.4.1 (06/02/03)
-----

* Fixed FPE caused by division by zero in the track time code.

0.4
---

* Now returns current/total track time.
* Can seek to a time offset in track.

0.3
---

* emphasis() function now returns correct value.
* PCM conversion works on big-endian architectures.

0.2
---

* Using distutils to configure and build pymad.
* py_madfile_new can now be passed a file object as well as a filename.
* frame info can be retrieved from the MadFile object
* Numerous bugfixes (see ChangeLog for details)

0.1
---

* Initial release.
  - Can open an mp3 file and decode the stream.


test/test.py
#!/usr/bin/python

import glob
import os.path
import sys
import urllib

import ao

for p in glob.glob("build/lib.*"):
    sys.path.insert(0, p)

print sys.path

import mad

def play(u):
    mf = mad.MadFile(u)
    
    if mf.layer() == mad.LAYER_I:
        print "MPEG Layer I"
    elif mf.layer() == mad.LAYER_II:
        print "MPEG Layer II"
    elif mf.layer() == mad.LAYER_III:
        print "MPEG Layer III"
    else:
        print "unexpected layer value"
        
    if mf.mode() == mad.MODE_SINGLE_CHANNEL:
        print "single channel"
    elif mf.mode() == mad.MODE_DUAL_CHANNEL:
        print "dual channel"
    elif mf.mode() == mad.MODE_JOINT_STEREO:
        print "joint (MS/intensity) stereo"
    elif mf.mode() == mad.MODE_STEREO:
        print "normal L/R stereo"
    else:
        print "unexpected mode value"
            
    if mf.emphasis() == mad.EMPHASIS_NONE:
        print "no emphasis"
    elif mf.emphasis() == mad.EMPHASIS_50_15_US:
        print "50/15us emphasis"
    elif mf.emphasis() == mad.EMPHASIS_CCITT_J_17:
        print "CCITT J.17 emphasis"
    else:
        print "unexpected emphasis value"
        
    print "bitrate %lu bps" % mf.bitrate()
    print "samplerate %d Hz" % mf.samplerate()
    sys.stdout.flush()
    #millis = mf.total_time()
    #secs = millis / 1000
    #print "total time %d ms (%dm%2ds)" % (millis, secs / 60, secs % 60)
        
    dev = ao.AudioDevice(0, rate=mf.samplerate())
    while 1:
        buffy = mf.read()
        if buffy is None:
            break
        dev.play(buffy, len(buffy))
        #print "current time: %d ms" % mf.current_time()

if __name__ == "__main__":
    print "pymad version %s" % mad.__version__
    for file in sys.argv[1:]:
        u = urllib.urlopen(file)
        if u:
        #if os.path.exists(file):
            print "playing %s" % file
            play(u)


test/madradio.py
#! /usr/bin/env python

import glob
import os.path
import socket
import sys
import urlparse

import ao

for p in glob.glob("build/lib.*"):
    sys.path.insert(0, p)

import mad


def madradio(url):
    scheme, netloc, path, params, query, fragment = urlparse.urlparse(url)
    try:
        host, port = netloc.split(':')
    except ValueError:
        host, port = netloc, 80
    if not path:
        path = '/'
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((host, int(port)))
    sock.send('GET %s HTTP/1.0\r\n\r\n' % path)
    reply = sock.recv(1500)
    #print repr(reply)
    file = sock.makefile()
    mf = mad.MadFile(file)
    print "bitrate %lu bps" % mf.bitrate()
    print "samplerate %d Hz" % mf.samplerate()
    dev = ao.AudioDevice(0, rate=mf.samplerate())
    while True:
        buffy = mf.read()
        if buffy is None:
            break
        dev.play(buffy, len(buffy))

if __name__ == '__main__':
    import sys
    try:
        url = sys.argv[1]
    except IndexError:
        #url = 'http://62.67.195.6:8000' # lounge-radio.com
        #url = 'http://63.241.4.18:8069' # xtcradio.com
        url = 'http://mp2.somafm.com:2040' # somafm
    madradio(url)


src/xing.c
/*
 * mad - MPEG audio decoder
 * Copyright (C) 2000-2001 Robert Leslie
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id: xing.c,v 1.1 2003/01/11 13:16:08 jaq Exp $
 */

#include "xing.h"
#include "mad.h"

#define XING_MAGIC (('X' << 24) | ('i' << 16) | ('n' << 8) | 'g')

/*
 * NAME:        xing->init()
 * DESCRIPTION: initialize Xing structure
 */
void xing_init(struct xing *xing) { xing->flags = 0; }

/*
 * NAME:        xing->parse()
 * DESCRIPTION: parse a Xing VBR header
 */
int xing_parse(struct xing *xing, struct mad_bitptr ptr, unsigned int bitlen) {
  if (bitlen < 64 || mad_bit_read(&ptr, 32) != XING_MAGIC)
    goto fail;

  xing->flags = mad_bit_read(&ptr, 32);
  bitlen -= 64;

  if (xing->flags & XING_FRAMES) {
    if (bitlen < 32)
      goto fail;

    xing->frames = mad_bit_read(&ptr, 32);
    bitlen -= 32;
  }

  if (xing->flags & XING_BYTES) {
    if (bitlen < 32)
      goto fail;

    xing->bytes = mad_bit_read(&ptr, 32);
    bitlen -= 32;
  }

  if (xing->flags & XING_TOC) {
    int i;

    if (bitlen < 800)
      goto fail;

    for (i = 0; i < 100; ++i)
      xing->toc[i] = mad_bit_read(&ptr, 8);

    bitlen -= 800;
  }

  if (xing->flags & XING_SCALE) {
    if (bitlen < 32)
      goto fail;

    xing->scale = mad_bit_read(&ptr, 32);
    bitlen -= 32;
  }

  return 0;

fail:
  xing->flags = 0;
  return -1;
}


src/xing.h
/*
 * mad - MPEG audio decoder
 * Copyright (C) 2000-2001 Robert Leslie
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id: xing.h,v 1.1 2003/01/11 13:16:08 jaq Exp $
 */

#ifndef XING_H
#define XING_H

#include "mad.h"

struct xing {
  long flags;             /* valid fields (see below) */
  unsigned long frames;   /* total number of frames */
  unsigned long bytes;    /* total number of bytes */
  unsigned char toc[100]; /* 100-point seek table */
  long scale;             /* ?? */
};

enum {
  XING_FRAMES = 0x00000001L,
  XING_BYTES = 0x00000002L,
  XING_TOC = 0x00000004L,
  XING_SCALE = 0x00000008L
};

void xing_init(struct xing *);

#define xing_finish(xing) /* nothing */

int xing_parse(struct xing *, struct mad_bitptr, unsigned int);

#endif


src/pymadfile.h
/* $Id: pymadfile.h,v 1.5 2003/01/11 13:15:39 jaq Exp $
 *
 * python interface to libmad (the mpeg audio decoder library)
 *
 * Copyright (c) 2002 Jamie Wilkinson
 *
 * This program is free software, you may copy and/or modify as per
 * the GNU General Public License (version 2, or at your discretion,
 * any later version).  This is the same license as libmad.
 */

#ifndef __PY_MADFILE_H__
#define __PY_MADFILE_H__

#include <Python.h>
#include <mad.h>

typedef struct {
  PyObject_HEAD PyObject *fobject;
  int close_file;
  struct mad_stream stream;
  struct mad_frame frame;
  struct mad_synth synth;
  mad_timer_t timer;
  unsigned char *buffy;
  unsigned int bufsiz;
  unsigned int framecount;
  unsigned long total_length;
} py_madfile; /* MadFile */

#define PY_MADFILE(x) ((py_madfile *)x)
#define MAD_STREAM(x) (PY_MADFILE(x)->stream)
#define MAD_FRAME(x) (PY_MADFILE(x)->frame)
#define MAD_SYNTH(x) (PY_MADFILE(x)->synth)
#define MAD_BUFFY(x) (PY_MADFILE(x)->buffy)
#define MAD_BUFSIZ(x) (PY_MADFILE(x)->bufsiz)
#define MAD_TIMER(x) (PY_MADFILE(x)->timer)

extern PyTypeObject py_madfile_t;

static void py_madfile_dealloc(PyObject *self, PyObject *args);
static PyObject *py_madfile_read(PyObject *self, PyObject *args);
static PyObject *py_madfile_layer(PyObject *self, PyObject *args);
static PyObject *py_madfile_mode(PyObject *self, PyObject *args);
static PyObject *py_madfile_samplerate(PyObject *self, PyObject *args);
static PyObject *py_madfile_bitrate(PyObject *self, PyObject *args);
static PyObject *py_madfile_emphasis(PyObject *self, PyObject *args);
static PyObject *py_madfile_total_time(PyObject *self, PyObject *args);
static PyObject *py_madfile_current_time(PyObject *self, PyObject *args);
static PyObject *py_madfile_seek_time(PyObject *self, PyObject *args);
static PyObject *py_madfile_getattr(PyObject *self, char *name);

#endif /* __PY_MADFILE_H__ */


src/madmodule.c
/* $Id: madmodule.c,v 1.4 2002/08/18 06:21:48 jaq Exp $
 *
 * python interface to libmad (the mpeg audio decoder library)
 *
 * Copyright (c) 2002 Jamie Wilkinson
 *
 * This program is free software, you may copy and/or modify as per
 * the GNU General Public License (version 2, or at your discretion,
 * any later version).  This is the same license as libmad.
 */

#include <mad.h>
#include "madmodule.h"

static PyMethodDef mad_methods[] = {
    {"MadFile", py_madfile_new, METH_VARARGS, ""}, {NULL, 0, 0, NULL}};

/* this handy tool for passing C constants to Python-land from
 * http://starship.python.net/crew/arcege/extwriting/pyext.html
 */
#define PY_CONST(x) PyDict_SetItemString(dict, #x, PyInt_FromLong(MAD_##x))

void initmad(void) {
  PyObject *module, *dict;

  module = Py_InitModule("mad", mad_methods);
  dict = PyModule_GetDict(module);

  PyDict_SetItemString(dict, "__version__", PyString_FromString(VERSION));

  /* layer */
  PY_CONST(LAYER_I);
  PY_CONST(LAYER_II);
  PY_CONST(LAYER_III);

  /* mode */
  PY_CONST(MODE_SINGLE_CHANNEL);
  PY_CONST(MODE_DUAL_CHANNEL);
  PY_CONST(MODE_JOINT_STEREO);
  PY_CONST(MODE_STEREO);

  /* emphasis */
  PY_CONST(EMPHASIS_NONE);
  PY_CONST(EMPHASIS_50_15_US);
  PY_CONST(EMPHASIS_CCITT_J_17);

  if (PyErr_Occurred())
    PyErr_SetString(PyExc_ImportError, "mad: init failed");
}
